extensions/amp-video-docking/1.0/dock.js

```diff
@@ -0,0 +1,488 @@
+          dockedStyle: dict({
+            ...transformToAreaStyle,
+            ...transitionTiming,
+            'transition-property': 'transform',
+          }),
+          placeholderIcon: dict({
+            ...fadeStyle,
+            ...transitionTiming,
+            'transform': transform(placeholderIconX, 0, 1),
+          }),
+        });
+      })
+      .then(
+        () =>
+          new Promise((resolve) =>
+            setTimeout(() => {
+              resolve(true);
+              this.isInTransition_ = false;
+              this.setState_({isInTransition: false});
+            }, transitionDurationMs)
+          )
+      );
+  }
+
+  /**
+   * @param {!IntersectionObserverEntry} entry
+   * @param {!DockTargetDef} target
+   * @param {number=} step
+   * @return {{x: number, y: number, scale: number, relativeX: !DirectionX}}
+   * @private
+   */
+  getDims_(entry, target, step) {
+    return interpolatedBoxesTransform(
+      entry.intersectionRect,
+      target.rect,
+      step
+    );
+  }
+
+  /** @private */
+  undock_() {
+    if (!this.current_ || this.isInTransition_) {
+      return;
+    }
+    const step = 0;
+    const {entry, target} = this.current_;
+    const {x, y, scale, relativeX} = this.getDims_(entry, target, step);
+
+    // Do not animate transition if video is out-of-view. Chrome glitches
+    // when pausing an out-of-view video, so we need to show controls and
+    // transition after. If we animate the transition, we would see native
+    // controls during, which looks a bit funky.
+    // const transitionDurationMs = isMostlyInView
+    //   ? this.calculateTransitionDuration_(step)
+    //   : 0;
+    // TODO(alanorozco): Unconfirmed on Preact version.
+
+    const transitionDurationMs = 300;
+
+    return this.placeAt_(
+      entry,
+      x,
+      y,
+      scale,
+      step,
+      transitionDurationMs,
+      relativeX,
+      'absolute'
+    ).then(() => {
+      this.current_ = null;
+      this.setStyles_(null);
+    });
+  }
+
+  /**
+   * Returns the area's target when a video should be docked.
+   * @param {!IntersectionObserverEntry} entry
+   * @return {?DockTargetDef}
+   * @private
+   */
+  getTargetFor_(entry) {
+    if (
+      this.current_
+      // this.isDragging_ ||
+      // !this.isValidSize_(video) ||
+      // this.ignoreBecauseAnotherDocked_(video) ||
+      // this.ignoreDueToNotPlayingManually_(video)
+    ) {
+      return null;
+    }
+
+    const {intersectionRect} = entry;
+    if (!isSizedRect(intersectionRect)) {
+      return null;
+    }
+    if (intersectionRect.top > this.getTopBoundary_()) {
+      return null;
+    }
+    return this.getUsableTarget_(entry);
+  }
+
+  /**
+   * @return {number}
+   * @private
+   */
+  getTopBoundary_() {
+    // TODO: Not supporting slot yet.
+    // const slot = this.getSlot_();
+    // if (slot) {
+    //   // Match slot's top edge to tie transition to element.
+    //   return slot.getPageLayoutBox().top;
+    // }
```

Comment:
Optional "slots" (not in the shadow DOM sense) are [invisible elements whose position determine the dock area](https://amp.dev/documentation/components/amp-video-docking/#custom-position-by-%22slot%22). I did not implement this since I'm not sure what the best way to do this should be—my gut is that `<Dock>` should receive a `DOMRect` which the AMP version should get from an element.